org/mozilla/jss/ssl/javax/JSSParameters.java

```diff
@@ -54,6 +57,11 @@ public JSSParameters(SSLParameters downcast) {
         if (downcast.getNeedClientAuth()) {
             setNeedClientAuth(downcast.getNeedClientAuth());
         }
+
+        String[] alpn = JDKCompat.SSLParametersHelper.getApplicationProtocols(downcast);
+        if (alpn != null) {
+            setApplicationProtocols(alpn);
+        }
```

Comment:
This is how can end up with *empty*, non-null list of protocols.

Reply:
Sorry, how?

`appProtocols` by default is initialized to `null`.
If `alpn` is `null`, we leave the default.
If `alpn` is non-null, we continue to the string version of `setApplicationProtocols`.

`setApplicationProtocols(String[] protocols)` sets `appProtocols` to `null` if `protocols` is `null`. If not, we build a byte array (checking lengths and ensuring no interior protocol is `null`), and pass it to  `setApplicationProtocols(byte[][] protocols)`.

This last method sets `appProtocols` to `null` if `protocols` is `null` or has zero length. This prevents us from setting an empty array (0 length array) as a non-`null` `appProtocols` list.

Otherwise, we validate lengths, ensure no interior member is `null` and then construct the result as a new `byte[][] appProtocols`.

Reply:
> Sorry, how?

This `JSSParameters` constructor takes an `SSLParameters`.  The OpenJDK `SSLParameters` class initialises the application protocols to empty array, not null.  See
https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/77b682e2d679/src/share/classes/javax/net/ssl/SSLParameters.java#l88 .

So aside from good hygiene, we really do need to handle the `(alpn != null && alpn.length < 1)` case.

Reply:
Like I said, this is already handled by `setApplicationProtocols(byte[][] protocols)`. The following code passes on `9c3f195b3114a5c4a1c33653ada7768f1af67987` (You can see above when I pushed this 5 days ago).

What commit are you using?

I pushed an explicit test case on top of it in `a6c2f8ccf57f7fa1caec940944918ed7803fb5ce`. That's the only commit I introduced today.

This is already handled.

Reply:
Thanks, you are right, I missed that the fix was introduced with the `byte[]`-based method commit.